experiment/theory.md

```diff
@@ -1 +1,70 @@
+
+**Pseudo Code**
+
+    BellmanFord(Graph, source):
+        distance = [] 
+        distance[source] = 0
+        for each vertex v in Graph: 
+            if v != source
+                distance[v] = infinity
+        for i = 1 to size of Graph:
+            for each edge (u, v) with weight w in Graph:
+                if distance[u] + w < distance[v]:
+                    distance[v] = distance[u] + w
+    
+    In the above algorithm distance[v] represents the length of the shortest path from source to v.
+    They are initialized to infinity except for the source vertex, which is initialized to 0.
+    The outer loop is repeated V-1 times, where V is the number of vertices in the graph.
+    The inner loop relaxes all the edges in the graph.
+
+**Time Complexity**
+
+From the above pseudo code we can see that the outer loop is repeated V-1 times and the inner loop is repeated E times where E is the number of edges in the graph. So the time complexity of the algorithm is O(V*E). This is the average time complexity of the algorithm. In the worst case we would have a complete graph then E would be equal to V^2. So the time complexity of the algorithm in the worst case would be O(V^3). We can optimize the algorithm by stooping the outer loop if we have not updated the distance of any vertex in the inner loop. Thus our best case time complxity will be equal to O(E).
+
+**Space Complexity**
+
+The space complexity of the algorithm is O(V) as we are using an array to store the distance of all the vertices from the source vertex. In the above pseudo code we are only calculating the shortest distance but the shortest path can be calculated by storing the parent of each vertex in an array. This will increase the space complexity of the algorithm to O(V*2) beacuse we are using two arrays to store the distance and parent of each vertex.
+
+
+**Pseudo Code for optimized algorithm with pareent array**
+
+    BellmanFord(Graph, source):
+        distance = [] 
+        parent = []
+        distance[source] = 0
+        for each vertex v in Graph: 
+            if v != source
+                distance[v] = infinity
+                parent[v] = null
+        for i = 1 to size of Graph:
+            for each edge (u, v) with weight w in Graph:
+                if distance[u] + w < distance[v]:
+                    distance[v] = distance[u] + w
+                    parent[v] = u
+
+
+**Negative Edges and Negative Cycle**
+
+Bellman-Ford is used over Dijkstra's Algorithm beacuse it handles garph with negative edges, but we need to ensure that the graph doesn't contain any negative cycle. In case of negative cycle the shortest distance will be minus infinity. For ex conider a cycle between edges A,B,C. The weights are 
```

Comment:
Graph misspelled as garph

For example, consider a cycle
in place of
For ex conider a cycle

Reply:
corrected these mistakes